File: 200_Number_of_Islands.py
class Solution(object):
    def numIslands(self, grid):
        m, n = len(grid), len(grid[0])
        isles = 0

        self.dr = [-1, 1, 0, 0]
        self.dc = [0, 0, -1, 1]

        for i in range(m):
            for j in range(n):
                if grid[i][j] == '1':
                    isles += 1
                    self.dfs(i, j, m, n, grid)

        return isles

    def dfs(self, i, j, m, n, grid):
        if i < 0 or j < 0 or i == m or j == n:
            return
        if grid[i][j] == '0':
            return

        grid[i][j] = '0'

        for k in range(4):
            self.dfs(i + self.dr[k], j + self.dc[k], m, n, grid)

File: cpp/200_Number_of_Islands.cpp
class Solution {
public:
    int dr[4] = {-1, 1, 0, 0};
    int dc[4] = {0, 0, 1, -1};
    void dfs(int i, int j, int m, int n, vector<vector<char>>& grid)
    {
        if (i < 0 || j < 0 || i == m || j == n) return;
        if (grid[i][j] == '0') return;

        grid[i][j] = '0';
        
        for (int k = 0; k < 4; ++k)
            dfs(i + dr[k], j + dc[k], m, n, grid);
    }
    int numIslands(vector<vector<char>>& grid) {
        int m = grid.size(), n = grid[0].size();

        int islands = 0;
        for (int i = 0; i < m; ++i)
        {
            for (int j = 0; j < n; ++j)
            {
                if (grid[i][j] == '1')
                {
                    ++islands;
                    dfs(i, j, m, n, grid);
                }
            }
        }

        return islands;
    }
};
